### ISSUE TITLE
mAP for test set.

### ISSUE THREAD

**Issue by sumbalakram** [ultralytics/ultralytics#3920]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I need to know that how the mAP has been calculated for whole dataset. As I want to calculate it for my custom test set.

**Comment by glenn-jocher** [comment]
@sumbalakram hi there!

The mAP (mean Average Precision) calculation in Ultralytics YOLOv8 is achieved by evaluating the model's performance across all classes in your dataset. It is computed during the Validation step, after each epoch during training as well as when you run the model in the 'Val' mode on a trained model.

When calculating the mAP for a custom test set, it's important to make sure your test set is correctly prepared and represented in your dataset YAML file. When using the 'Val' mode on your test set, it calculates predictions for each image, then compares these predictions to the true values. The model outputs a Precision and Recall value for each class, and the mAP is the average of these Precision values across all classes.

Do ensure that your ground truth bounding boxes in the test set are correctly labeled as YOLOv8 uses this information for computing the mAP. Each bounding box label requires the class, as well as the normalized x, y (center of box), width and height values.

Please do remember that the mAP itself is a summary metric of Precision and Recall across all your object classes, and may not fully capture the nuances of your model's performance on the test data. It can be beneficial to also look at class specific Precision and Recall, or even individual detection results to get a complete understanding of your model's performance.

Best of Luck with your work!

**Comment by TuanAnh24402** [comment]
Hi Glenn, i want to caculate mAP in testing in YOLOv8. But I don't know that. Can you help me?

**Comment by glenn-jocher** [comment]
Hi there! 😊 Sure, I'd be happy to guide you on calculating mAP for testing in YOLOv8. If you're using a custom test set, ensure it's correctly defined in your dataset YAML file. Then, you can simply use the `val` mode with your test set like this:

```bash
yolo val model=yolov8n.pt data=your_test_data.yaml
```

This will run the model on your test data and calculate metrics, including mAP. Make sure your test data annotations are in the same format as your training data. Good luck, and feel free to ask if you have more questions!

**Comment by TuanAnh24402** [comment]
@glenn-jocher Hi Glenn. How do I edit the yaml file? To be able to calculate mAP for testing. If I remove the 2 folders train and val, the program will report an error "You must have all 3 folders". I am sorry my english skill is not good. Thanks for help me.
This is my file yaml:
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 36
names: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']

**Comment by glenn-jocher** [comment]
Hi there! 😊 It looks like you're on the right track with editing the YAML file. To calculate mAP for your test dataset without needing to provide train and val paths, you just need to ensure `test` path is specified as you have done.

If your use case only involves evaluating the model on the test set, and you won't be training or validating, you can simply comment out the `train` and `val` paths in your YAML like this:

```yaml
#train: ../train/images
#val: ../valid/images
test: ../test/images

nc: 36
names: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
```

Then, you can run the validation command pointing it to use your test data:

```bash
yolo val model=yolov8n.pt data=your_yaml_file.yaml
```

This should work just fine! Do let me know if there's anything else you need help with. Good luck! 😊

**Comment by TuanAnh24402** [comment]
After i had trained model yolov8. I did not see parameters confusion matrix.
![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/151445372/7f4ab783-45f3-4b45-9fa7-9cf1455f4ecc)

**Comment by TuanAnh24402** [comment]
After I trained model yolov8. I did not see parameters confusion matrix.
![confusion_matrix_test](https://github.com/ultralytics/ultralytics/assets/151445372/dab2513e-537b-4444-b19d-ad29515cee9b)

**Comment by glenn-jocher** [comment]
Hi there! It looks like you may need to use the `--conf-mat` argument during validation to generate a confusion matrix. Here's how you can do it:

```bash
yolo val model=yolov8n.pt data=your_dataset.yaml --conf-mat
```

This should output the confusion matrix for your model. Let me know if there's anything else you need help with! 😊

**Comment by TuanAnh24402** [comment]
Hi Glenn. I don't understand "yolo val model=yolov8n.pt data=your_dataset.yaml --conf-mat". What should I do with this code? Can you guide me in more detail?

**Comment by glenn-jocher** [comment]
@TuanAnh24402 hello! Absolutely, I'd be happy to clarify that for you.

The command `yolo val model=yolov8n.pt data=your_dataset.yaml --conf-mat` is used to validate your trained YOLOv8 model and get a confusion matrix which helps in understanding how well your model is performing across different classes.

Here's a step-by-step breakdown of what you need to do:
1. Ensure your `yolov8n.pt` (your trained model file) and `your_dataset.yaml` (this should include the path to your validation dataset) are in the correct locations.
2. Open your terminal.
3. Navigate to the directory where your model and dataset YAML file are located.
4. Run the command provided.

This will start the validation process and generate a confusion matrix which will be saved in your runs folder.

Let me know if you need further assistance! 😊
